### PR TITLE
Add Support for Dynamic Asset Collections

### PR DESCRIPTION
**Pull request by lgrossi**
This PR introduces the ability to load and build collections of dynamic assets in bevy_asset_loader, addressing the need outlined in issue [#186](https://github.com/NiklasEi/bevy_asset_loader/issues/186). The enhancement allows for more flexible asset management, enabling users to define and group dynamic assets in structured collections.

## Motivation
Previously, loading multiple dynamic assets required individual handling for each asset, complicating the asset loading process and cluttering the codebase. With this update, users can now group related assets together, simplifying the loading process (specially for applications that handles a huge number of assets dynamically).

## Changes
DynamicAsset Trait for Vec<StandardDynamicAsset>: Implemented the DynamicAsset trait for Vec<StandardDynamicAsset>, enabling the loading and building of asset collections.

StandardDynamicAssetArrayCollection: Introduced a new struct to represent collections of dynamic assets, allowing assets to be grouped and registered under a common key.

## Usage Example
With these changes, users can now define array of asset collections in RON format:
```ron
({
    "layouts": [
        TextureAtlasLayout (
            tile_size_x: 32.,
            tile_size_y: 32.,
            columns: 12,
            rows: 12,
        ),
        TextureAtlasLayout (
            tile_size_x: 32.,
            tile_size_y: 64.,
            columns: 12,
            rows: 6,
        ),
        TextureAtlasLayout (
            tile_size_x: 64.,
            tile_size_y: 32.,
            columns: 6,
            rows: 12,
        ),
        TextureAtlasLayout (
            tile_size_x: 64.,
            tile_size_y: 64.,
            columns: 6,
            rows: 6,
        ),
    ],
    "mixed": [
        StandardMaterial (
            path: "images/tree.png",
        ),
        Image (
            path: "ryot_mascot.png",
            sampler: Nearest
        ),
        Image (
            path: "ryot_mascot.png",
            sampler: Nearest
        ),
    ],
})
```
 
And the rust code would look like:
```rust
    #[asset(key = "layouts", collection(typed))]
    atlas_layout: Vec<Handle<TextureAtlasLayout>>,

    #[asset(key = "mixed", collection)]
    mixed_handlers: Vec<UntypedHandle>,
```

## Conclusion
This update offers a straightforward method for managing multiple dynamic assets within the bevy_asset_loader. It's particularly beneficial for applications dealing with a large number of assets dynamically, aiming to simplify the process and improve developer experience. By grouping assets into collections, it hopes to make dynamic asset handling more accessible and less cumbersome.

### CODE REVIEW

**Comment by NiklasEi** [comment]
Thanks for offering your solution :slightly_smiling_face: 
#151 would allow deserializing collections of dynamic assets without a new `DynamicAssetCollection` implementation but it has been blocked on the next `ron` release for a while. For this solution, users will have to register this new dynamic asset collection type with the loading state and cannot mix collections and single dynamic assets in the same file.
I would be open to merging this as a temporary feature until we can get #151 as the more general solution to the same issue. What do you think?

**Comment by lgrossi** [comment]
Hey, indeed #151 seems like a more general solution. Curious to know the reasons behind opting to not allowing single collections and multi collections in the same schema.

I'm happy with this being a temp solution til we reach there, but I'm also ok with this not being merged (I can use my temp branch til we got #151 merged). I was just aiming to share this since it's quite a effortless/clean work around for the time being.

Thanks for the inputs. :smile:

**Comment by NiklasEi** [comment]
> Curious to know the reasons behind opting for not allowing single collections and multi collections in the same schema.

I meant this PR doesn't allow it, because a single file will be loaded as `StandardDynamicAssetArrayCollection` or as `StandardDynamicAssetCollection`. #151 allows for mixing single assets and collections in the same file.

**Comment by lgrossi** [comment]
@NiklasEi I finally had some time to add documentation and tests. Please let me know if you want me to add anything else before merging this PR.

**Comment by lgrossi** [comment]
> @NiklasEi I finally had some time to add documentation and tests. Please let me know if you want me to add anything else before merging this PR.

Ok, there was some flakiness in the test due to the nature of RON serialization/deserialization (it doesn't guarantee the order of the elements). I changed it to compare before/after deserialized objects instead of the RON strings, it should be fine now.

File: bevy_asset_loader/src/standard_dynamic_asset.rs
Comment (NiklasEi) [inline bevy_asset_loader/src/standard_dynamic_asset.rs:390]: We might want to add this type to the prelude

**Review by NiklasEi — APPROVED** [review]

**Comment by lgrossi** [comment]
Hi @NiklasEi is there anything else I can do to help getting this merged? Let me know if I can help :smile:

**Pull request merged**
